Adding jHyphen as a fallback

kalaspuffar opened the pull request:
As the current implementation in dotify doesn't support all daisy organization languages we decided to implement a fallback in order to run hyphenation with libhyphen if the language isn't supported by the built-in latex hyphenator.

kalaspuffar committed:
First implementation.

kalaspuffar committed:
Add OS check.

kalaspuffar committed:
Make latex default.

kalaspuffar committed:
More error information.

kalaspuffar committed:
add libhyphen packages.

kalaspuffar committed:
Add sudo.

kalaspuffar committed:
Testing with sudo.

kalaspuffar commented:
Hi @bertfrees and @PaulRambags

I know that you don't use this code in the Pipeline but you might want to review it and give some feedback anyway.

Best regards
Daniel

kalaspuffar committed:
Testing with myspell instead.

kalaspuffar committed:
Update apt.

bertfrees commented on src/org/daisy/dotify/hyphenator/impl/JHyphenator.java:
```diff
@@ -0,0 +1,55 @@
+
+	@Override
+	public String hyphenate(String phrase) {
+		return ensureStartAndEnd(instance.hyphenate(phrase, SHY, ZWSP));
```
I believe this will throw an exception if the word contains non-standard hyphenation points. You might want to handle that?

bertfrees commented on src/org/daisy/dotify/hyphenator/impl/JHyphenator.java:
```diff
@@ -0,0 +1,55 @@
+
+    private String handleWord(String s) {
+
+	    if(s.length() <= getBeginLimit()) return s.replaceAll("" + SHY, "");
```
`replace` instead of `replaceAll` might be more efficient? Dunno.

bertfrees commented:
If you're interested I have libhyphen binaries for Windows and MacOS.

PaulRambags commented on src/org/daisy/dotify/hyphenator/impl/JHyphenator.java:
```diff
@@ -0,0 +1,55 @@
+        rest = s.substring(0, endIndex);
+        String end = s.substring(endIndex);
+
+	    return rest + end.replaceAll("" + SHY, "");
```
this line contains a mixture of TAB and space characters

PaulRambags commented on src/org/daisy/dotify/hyphenator/impl/JHyphenator.java:
```diff
@@ -0,0 +1,55 @@
+    private Hyphenator instance;
+
+	JHyphenator(String locale) throws HyphenatorConfigurationException {
+	    try {
```
this line contains a mixture of TAB and space characters

PaulRambags commented on src/org/daisy/dotify/hyphenator/impl/JHyphenator.java:
```diff
@@ -0,0 +1,55 @@
+	    String rest = s.substring(getBeginLimit());
+	    s = start.replaceAll("" + SHY, "") + rest;
+
+        if(s.length() <= getEndLimit()) return s.replaceAll("" + SHY, "");
```
You are performing the same action twice because `s` consists of `start` + `rest` and `start` was already processed in the previous statement. That is not so nice.

kalaspuffar commented:
Hi, @bertfrees and @PaulRambags.

I've rewritten the code a bit to improve performance and hopefully resolve some of the issues you've reported. Not much of the old code is left :)

So if you have time, please review it again.

PS. @bertfrees I've also tested a bit more, and jhyphen has pretty much the same performance if we use it as I implemented it here. I think most of the issues with time can be because of the regression test. I think they take longer if many documents differ. DS.

Best regards
Daniel

kalaspuffar committed:
Rewrote for performance.

bertfrees commented:
Looks good.

> I think most of the issues with time can be because of the regression test. I think they take longer if many documents differ.

Ah yes, that could be an explanation.

kalaspuffar commented:
Hi @bertfrees 

Does that mean that you approve this PR as well? When it comes to the other DLLs or so files for the jhyphen library, then we are interested in getting hold of these. We would like to test it on all systems, at least. 

Do you have any code to locate the dictionary files on different platforms?

Best regards
Daniel

kalaspuffar committed:
Changed begin and end to be one less.

bertfrees commented:
You don't need my approval. But as far as I was able to tell (I didn't look into every little aspect of the change) it looked OK.

In my code I embed the binaries and install them in some location before invoking them so I don't need to locate them. I don't have a fallback to use system libraries. You can find the binaries here: https://repo1.maven.org/maven2/org/daisy/libs/hyphen/2.8.8/. They are a bit hidden inside the JARs but you'll find them. There are versions for Linux, Windows and Mac OS, and for both 32 and 64 bit architectures.
